File: app/src/main/java/it/ibashkimi/lockscheduler/model/LockAction.java
package it.ibashkimi.lockscheduler.model;


public class LockAction extends Action {

    private LockMode lockMode;

    public LockAction() {
        this(new LockMode(LockMode.LockType.UNCHANGED));
    }

    public LockAction(LockMode lockMode) {
        super(Type.LOCK);
        this.lockMode = lockMode;
    }

    public LockMode getLockMode() {
        return lockMode;
    }

    public void setLockMode(LockMode lockMode) {
        this.lockMode = lockMode;
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof LockAction)) {
            return false;
        }
        LockAction action = (LockAction) obj;
        return getLockMode().equals(action.getLockMode());
    }
}

File: app/src/main/java/com/ibashkimi/lockscheduler/model/action/LockAction.kt
package com.ibashkimi.lockscheduler.model.action

import android.os.Parcelable
import kotlinx.android.parcel.Parcelize

@Parcelize
data class LockAction(val lockMode: LockMode) : Action(Type.LOCK), Parcelable {

    enum class LockType(val value: String) {
        PIN("pin"), PASSWORD("password"), SWIPE("swipe"), UNCHANGED("unchanged")
    }

    sealed class LockMode(val lockType: LockType) : Parcelable {
        @Parcelize
        data class Pin(val input: String) : LockMode(LockType.PIN)

        @Parcelize
        data class Password(val input: String) : LockMode(LockType.PASSWORD)

        @Parcelize
        object Swipe : LockMode(LockType.SWIPE)

        @Parcelize
        object Unchanged : LockMode(LockType.UNCHANGED)
    }
}
